drivers/mpu9250/MPU9250.cpp

```diff
@@ -328,37 +415,109 @@ void MPU9250::_measure()
 		int16_t		gyro_x;
 		int16_t		gyro_y;
 		int16_t		gyro_z;
-	} report;
+		//uint8_t		ext_data[24];
+	};
 #pragma pack(pop)
 
-	_bulkRead(MPUREG_ACCEL_XOUT_H, (uint8_t *)&report, sizeof(report));
+	// Get FIFO byte count to read and floor it to the report size.
+	int bytes_to_read = get_fifo_count() / sizeof(int_status_report) * sizeof(int_status_report);
 
-	/* TODO: add ifdef for endianness */
-	report.accel_x = swap16(report.accel_x);
-	report.accel_y = swap16(report.accel_y);
-	report.accel_z = swap16(report.accel_z);
-	report.temp = swap16(report.temp);
-	report.gyro_x = swap16(report.gyro_x);
-	report.gyro_y = swap16(report.gyro_y);
-	report.gyro_z = swap16(report.gyro_z);
+	if (bytes_to_read < 0) {
+		// TODO: count error
+		return;
+	}
 
-	m_synchronize.lock();
+	// The FIFO buffer on the MPU is 512 bytes according to the datasheet, so let's use
+	// 36*14 = 504.
+	const unsigned buf_len = 36 * sizeof(int_status_report);
+	uint8_t fifo_read_buf[buf_len];
 
-	m_sensor_data.accel_m_s2_x = float(report.accel_x) * (MPU9250_ONE_G / 2048.0f);
-	m_sensor_data.accel_m_s2_y = float(report.accel_y) * (MPU9250_ONE_G / 2048.0f);
-	m_sensor_data.accel_m_s2_z = float(report.accel_z) * (MPU9250_ONE_G / 2048.0f);
-	m_sensor_data.temp_c = float(report.temp) / 361.0f + 35.0f;
-	m_sensor_data.gyro_rad_s_x = float(report.gyro_x) * GYRO_RAW_TO_RAD_S;
-	m_sensor_data.gyro_rad_s_y = float(report.gyro_y) * GYRO_RAW_TO_RAD_S;
-	m_sensor_data.gyro_rad_s_z = float(report.gyro_z) * GYRO_RAW_TO_RAD_S;
+	if (bytes_to_read <= 0) {
+		// TODO: increase error counter if there is nothing to read.
+		return;
+	}
 
-	m_sensor_data.last_read_time_usec = DriverFramework::offsetTime();
-	m_sensor_data.read_counter++;
+	const unsigned read_len = MIN((unsigned)bytes_to_read, buf_len);
 
-	_publish(m_sensor_data);
+	memset(fifo_read_buf, 0x0, buf_len);
 
-	m_synchronize.signal();
-	m_synchronize.unlock();
+	// According to the protocol specs, all sensor and interrupt registers may be read at 20 MHz.
+	// It is unclear what rate the FIFO register can be read at.
+	// If the FIFO buffer was read at 20 MHz, two effects were seen:
+	// - The buffer is off-by-one. So the report "starts" at &fifo_read_buf[i+1].
+	// - Also, the FIFO buffer seemed to prone to random corruption (or shifting), unless
+	//   all other sensors ran very smooth. (E.g. Changing the bus speed of the HMC5883 driver from
+	//   400 kHz to 100 kHz could cause corruption because this driver wouldn't run as regularly.
+	//
+	// Luckily 10 MHz seems to work fine.
+
+	_setBusFrequency(SPI_FREQUENCY_10MHZ);
+	result = _bulkRead(MPUREG_FIFO_R_W, fifo_read_buf, read_len);
+
+	if (result != 0) {
+		// TODO: increase error counter.
+		return;
+	}
+
+	for (unsigned i = 0; i + sizeof(int_status_report) < read_len; i += sizeof(int_status_report)) {
```

Comment:
`read_len` is a multiple of `sizeof(int_status_report)`, and _bulkRead seems to read as much as it gets told to.

This reminds me that I should rename `int_status_report` into `fifo_packet`.